### PR TITLE
Refactor the Makefiles to be more DRY/simple.

### PR DESCRIPTION
**Pull request by tstrass**
### Summary
Each measure still needs a Makefile, but they're much smaller and simpler now. The "master" Makefile in the root of the repo is the only one you need to use from the command line to perform patient generation and preload images. The master makes calls to the measure Makefiles.

The master makefile at a high level has two distinct functions: generating patients, and preloading CQF Ruler docker images. Here are some example commands:

Generate patients for EXM 130:
```
make MEASURE_DIR=EXM_130 r4
```
```
make MEASURE_DIR=EXM_130 stu3
```

Preload a docker image:
```
make MEASURE_DIR=EXM_130 preload-r4
```
```
make MEASURE_DIR=EXM_130 preload-stu3
```

We now only have one `connectathon` clone and one `synthea` clone for the whole repo, rather than one for each measure. Also, all of the "seeding" is consolidated into one place, and is performed with an "all or nothing" approach. So even if you're only generating patients for measure A, it will seed the image with all of the measures. I made this decision because it helps consolidate all of our references to `connectathon` resources, and there is only a marginal performance hit (since we're now performing a bunch of technically unnecessary PUT and POST calls).

We still need to have small Makefiles for each measure, which the master Makefile calls. This essentially boils down to the fact that each measure has it's own Synthea module and it's own CQL file. There is probably still some room for further simplification/DRYing in this area, but this seemed like a reasonable major first step.

### Things to note
- No matter the measure, this now outputs to the single Synthea clone that exists in the root of the repo. As such, if you are calculating patients for multiple measures in a row, you will need to clear that directory between actions. `make clean` clears this directory, but the other steps performed by the `clean` aren't necessary between calculations.

### Testing
The only way to really test this is to run a bunch of commands against the master Makefile, testing generating patients and preloading images with different measures, using commands like the examples I gave above.

### CODE REVIEW

**Review by lukoktonos — CHANGES_REQUESTED** [review]
Re-running the top level make command repeats the `.seed*` targets when run multiple times.  Additional target markers for `.run-load-script-stu3` and `.run-load-script-r4` may be needed.

For both r4 and stu3, I am getting this error on the calculation step:

```
Request failed with status code 405
    at processBundles (/usr/local/lib/node_modules/fhir-bundle-calculator/src/cli.js:112:13)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
(node:18241) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
```

File: .gitignore [hunk 1]
```diff
@@ -1,9 +1,9 @@
 output
 synthea
 connectathon
-.cqf-ruler
-.seed-cqf-ruler*
-.seed-cqf-ruler-no-vs*
+.new-cqf-ruler
```
Comment (losborne) [inline .gitignore]:
```suggestion
.new-cqf-ruler
.wait-cqf-ruler
```

**Comment by lukoktonos** [comment]
I went back to look into the 405s today and they no longer happen for me. Maybe restarting docker or my computer fixed something--could be related to heap memory issue we were seeing.

Now patients are calculating if I use a locally running cqf-ruler, which I think is what we are expecting.  So my only comment is to prevent the duplicate POSTs and PUTs when `make` is run more than once.

File: Makefile [hunk 2]
```diff
@@ -1,25 +1,204 @@
...
+		-H 'Content-Type: application/json' \
+		-d @connectathon/fhir4/bundles/EXM124_FHIR4-8.2.000/EXM124_FHIR4-8.2.000-files/measure-EXM124_FHIR4-8.2.000.json
+	curl -X POST http://localhost:8080/cqf-ruler-r4/fhir \
+		-H 'Content-Type: application/json' \
+		-d @connectathon/fhir4/bundles/EXM124_FHIR4-8.2.000/EXM124_FHIR4-8.2.000-files/library-deps-EXM124_FHIR4-8.2.000-bundle.json
+	curl -X PUT http://localhost:8080/cqf-ruler-r4/fhir/Library/library-EXM124-FHIR4-8.2.000 \
+  		-H 'Content-Type: application/json' \
+		-d @./connectathon/fhir4/bundles/EXM124_FHIR4-8.2.000/EXM124_FHIR4-8.2.000-files/library-EXM124_FHIR4-8.2.000.json
+	# CMS 125
+	curl -X PUT http://localhost:8080/cqf-ruler-r4/fhir/Measure/measure-EXM125_FHIR4-7.2.000 \
```
Comment (lukoktonos) [inline Makefile]:
```suggestion
	curl -X PUT http://localhost:8080/cqf-ruler-r4/fhir/Measure/measure-EXM125-FHIR4-7.2.000 \
```

**Comment by lukoktonos** [comment]
Right now the preload targets only preload 1 measure at a time, but I think we would only ever want to preload all measures? It may make sense to have that have all by default and perhaps allow for a single measure if desired.

Might also be useful to have a target that generates all patients in the same way rather than having to run once per measure.

File: EXM_105/Makefile [hunk 3]
```diff
@@ -1,56 +1,13 @@
...
-	cd synthea && ./run_synthea -a 18-21 -p $(PATIENT_COUNT) -m stroke_exm_105
-
-CQL_FILE := connectathon/fhir3/bundles/EXM105_FHIR3-8.0.000/EXM105_FHIR3-8.0.000-files/EXM105_FHIR3-8.0.000.cql
-calculate-patients:
-	calculate-bundles -d ./synthea/output/fhir_stu3 -c $(CQL_FILE) -u http://localhost:8080/cqf-ruler-dstu3/fhir --measure-id measure-EXM105-FHIR3-8.0.000
-
-clean:
-	-docker stop cqf-ruler
-	-rm -rf synthea output connectathon .cqf-ruler .seed-cqf-ruler .seed-cqf-ruler-no-vs .generate-patients .calculate-patients .setup-cqf-ruler
+generate-patients-stu3:
```
Comment (lukoktonos) [inline EXM_105/Makefile:7]: missing generate-patients-r4 for this Makefile for some reason (both before and after these changes)

**Comment by losborne** [comment]
I think currently in order to preload the cqf-ruler image these steps need to happen:

1. Load cqf-ruler with all resources (including valuesets), generate patients, and calculate patients for each measure separately
2. remove cqf-ruler instance that has valuesets
3. run. preload-r4 for each measure separately
4. run  .update-cqf-ruler-image

It may be possible to make all of those things happen with a single make command. Or at least one for steps 1-3 and have step 4 separate still.

**Comment by tstrass** [comment]
> Right now the preload targets only preload 1 measure at a time, but I think we would only ever want to preload all measures? It may make sense to have that have all by default and perhaps allow for a single measure if desired.
> 
> Might also be useful to have a target that generates all patients in the same way rather than having to run once per measure.

I agree, but since we didn't have that capability before, I would say that's out of the scope of this task. Will create a new task for that if you agree.

File: EXM_105/Makefile (hunk 3 above)
Comment (tstrass) [reply EXM_105/Makefile:7]: Yeah i guess we just haven't yet done R4 for this measure. Probably out of scope of this task as well though.

File: Makefile (hunk 2 above)
Comment (tstrass) [reply Makefile]: Huh I definitely combed though and copy-pasted all the IDs for every measure and library, but I guess that somehow didn't get committed. Anyway, they were all wrong not just this one so I just pushed a fix.

File: EXM_105/Makefile (hunk 3 above)
Comment (lukoktonos) [reply EXM_105/Makefile:7]: Sounds good--that will be a pre-req for the task to publish a new pre-loaded r4 docker image

File: Makefile (hunk 2 above)
Comment (lukoktonos) [reply Makefile]: Unfortunately the ids in the http requests should have hyphens, but the filenames have underscores.  Your commit changed the filenames, so they are wrong now.  I think the one I suggested was the only {id, filename} that was incorrect.

Comment (tstrass) [reply Makefile]: 🤦‍♂ welp that explains the mysterious "loss" of my changes

**Comment by mgramigna** [comment]
Question rather than a change request: do we think it would be useful to add a docker pull before running the cqf-ruler container to ensure that the image is up to date?

**Review by mgramigna — CHANGES_REQUESTED** [review]
I think the `<measure-dir>/stu3` and `<measure-dir>/r4` directories should be gitignored since they just contain the output from the bundle calc script

**Review by mgramigna — CHANGES_REQUESTED** [review]
Unfortunately, I get to a Heap space error really quickly. I think this should be avoidable if we don't seed cqf-ruler with every measure bundle when just running calculation for 1 measure

**Comment by lukoktonos** [comment]
> Unfortunately, I get to a Heap space error really quickly. I think this should be avoidable if we don't seed cqf-ruler with every measure bundle when just running calculation for 1 measure

What is your docker RAM set at?

**Comment by mgramigna** [comment]
@losborne it was at 2, bumping to 8 and trying again

**Comment by mgramigna** [comment]
I'm fine with heap space now.. Just need to update the IDs for things like Luke mentioned and this is looking pretty good

File: EXM_105/Makefile (hunk 3 above)
Comment (tstrass) [reply EXM_105/Makefile:7]: created https://jira.mitre.org/browse/TACOSTRAT-396

**Review by lukoktonos — APPROVED** [review]

**Review by mgramigna — APPROVED** [review]
L
G
T
M

**Pull request merged**